app/services/proxy_deployment_service.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+class ProxyDeploymentService
+  delegate :apicast_configuration_driven,
+           :deployable?,
+           :service_mesh_integration?,
+           :provider,
+           :proxy_configs, to: :@proxy
+
+  class UnknownStageError < ArgumentError; end
+
+  def self.call(*args)
+    new(*args).call
+  end
+
+  def initialize(proxy, environment: :staging, v1_compatible: false)
+    @proxy = proxy
+    @environment = environment
+    @v1_compatible = v1_compatible
+  end
+
+  def call
+    case @environment
+    when :staging, :sandbox
+      deploy
+    when :production
+      deploy_production
+    else
+      raise UnknownStageError, "Unknown environment: #{@stage}"
+    end
+  end
+
+  private
+
+  def deploy
+    return true unless deployable?
```

Comment:
The following makes more sense to me:

```suggestion
    return unless deployable?
```

Not being deployable does not mean that deployed it is, so returning `true` is confusing.

Reply:
Shouldn't we do that in another PR? One that affects master

Reply:
I think this is part of this PR as it introduces `ProxyDeploymentService`, but maybe that's is just me.

Reply:
But it's just a refactorization, it does not change any logic. By changing the return value.. it would change the behavior of current master.

Reply:
https://github.com/3scale/porta/pull/1470/files#diff-3ab35f0eaf168ec265ee8e446edc8ca8L633